### DISCUSSION TITLE
State conformance to `ObservableState` store runtime diagnostics

### DISCUSSION THREAD

**Discussion by maximkrouk** [pointfreeco/swift-composable-architecture#2682] [Beta]
Just got `SIGABRT` crash while playing with `observation-beta`

Can be reproduced in Example project [here](https://github.com/CaptureContext/combine-cocoa-navigation/tree/5c01ec3e52c078b7f7de04228206f354eaf548c6/Example).

Steps to reprocuce:
- Launch the app
- Open first user profile
- Go back
- Open first tweet details

Console error output:
```
Could not cast value of type 'Swift.Optional<()>' (0x1c1d89568) to 'ProfileFeature.ProfileFeature.State' (0x1032fe4e0).
```

Some notes from the callstack:
```
Store.send(_:originatingFrom:)
withoutPerceptionChecking(_:) // Kinda hints too look for missing `ObservableState`s, but it's not obvious at all
...
Store.init(rootStore:toState:fromAction).subscribeToDidSet()
PartialToState.keyPath.callAsFunction(_:) // on failure Xcode navigates to here (`return state❌[keyPath: keyPath] as! State`)
```

I already solved it by accidentally finding that I didn't mark `ProfileTabFeature.State` as `ObservableState`, but none of errors were helpful for finding whats wrong...

I think the MVP for the fix could be just a simple comment in `PartialToState.keyPath.callAsFunction(_:)` that "if you got a SIGABRT crash here, probably you forgot to add `@ObservableState` to one of your features"

Ideally it would be cool to use some failure with a message containing specific type that needs to conform to `ObservableState`, but I don't see how to do that without a bunch on reflection (which will probably drain performance, but probably it's ok for debug builds)

**Comment by mbrandonw** [comment]
Hi @maximkrouk, thanks for reporting this. Unfortunately this is due to Swift bug in key paths. We filed [this](https://github.com/apple/swift/issues/70611) yesterday and I just pushed a fix (60b5fe47c6) for the part that you were running into.

I was also able to verify that marking `FeedTabFeature.State` with `@ObservableState` did somehow fix the issue, but that is completely mystifying to me. FWIW, this crash should not happen at all, regardless of using `@ObservableState` or not. And so we don't think we need to create any documentation at the crash site. It's just a Swift bug that will hopefully be fixed sometime soon.

**Reply by maximkrouk** [reply]
> but that is completely mystifying to me

I find this part of the reply really hilarious 😂

> We filed https://github.com/apple/swift/issues/70611 yesterday and I just pushed a fix (https://github.com/pointfreeco/swift-composable-architecture/commit/60b5fe47c69f8a323375cb50ba4b928099b4acf5) for the part that you were running into.

👍

**Reply by mbrandonw** [reply]
Also something stood out in your demo is that you have some `SwitchStore`s, `ForEachStore`s, etc that can all be done away with using the new observation tools.

**Reply by maximkrouk** [reply]
I began to write the demo before updating to `observation-beta`, today I finished shaping the project and gonna start scaling it with new features and refactoring existing ones tomorrow 👌

But it's nice that having those Stores didn't affect migration 👍
